in-progress/8401-proof-timeliness/design.md

```diff
@@ -0,0 +1,289 @@
+# Proof Timeliness
+
+|                      |                                                             |
+| -------------------- | ----------------------------------------------------------- |
+| Issue                | https://github.com/AztecProtocol/aztec-packages/issues/8401 |
+| Owners               | @just-mitch                                                 |
+| Approvers            | @LHerskind                                                  |
+| Target Approval Date | 2024-09-13                                                  |
+
+
+## Executive Summary
+
+The rollup contract will require that the pending chain be pruned to the proven chain if a proof for an epoch does not get verified on L1 in a fixed amount of time.
+
+## Introduction
+
+There is functionality in the rollup contract to prune the pending chain, but this is not currently used.
+
+A strict proof timeliness requirement affords guarantees to users that their transactions will be included in the proven chain within a fixed amount of time.
+
+This design focuses on how those timelines are defined and enforced by the rollup contract; it does not cover how nodes will handle the ensuing reorg with respect to their local state.
+
+## Proving phases
+
+<!-- Editors: you can copy/paste the png from the repository into excalidraw to make edits. -->
+
+![Proving Phases](./proving-phases.png)
+
+### Proof claim phase
+
+The beginning of each epoch is the "proof claim phase".
+
+This phase has a duration of $C$ slots (e.g. 13 slots. See notebook for modeling).
+
+During this time, the proposer for a slot can submit a claim to be the prover for the previous epoch.
+
+This can be a transaction separate from the submission of their proposed block.
+
+Doing so grants monopoly rights to the rewards for submitting the proof of epoch `n-1`:
+if there are block rewards in test token TST, they will be sent to the address that submitted the claim.
+
+If no claim is submitted, the pending chain is pruned to the tip of the proven chain;
+a caveat is if a proof for epoch `n-1` is submitted during the proof claim phase.
+
+### Proof production
+
+The proof for epoch `n-1` must land in epoch `n`.
+
+Failure to submit a proof results in the bond posted with the claim being slashed if it exists.
```

Comment:
I think we need a "proof race" or some other method if the proof fails to arrive. Otherwise, the economic cost of rolling back a significant portion of the chain is "just" the prover bond, and we go back to the old "option" issue we had with Fernet before introducing the validator set: a prover (or a proposer, if they submit the prover bond themselves) can decide whether to roll back about 20 minutes of the chain at only the cost of the bond.

Reply:
That said, this is probably good enough for testnet.

Reply:
I think it is a nice result to force proofs of epoch `n-1` landing in `n` (instead of possibly in `n+1`), so to have an option for a proof race while keeping this property we would likely need to extend the length of the epochs and/or the duration of an L2 slot.

For example, it is possible to have a proof race period if we bump the number of slots in an epoch to 48, and the duration of a slot is 36 seconds, that puts us at 21 minutes after the proof claim period (i.e. enough time for the first proof to not land, and have enough time to make another), and total epoch length of ~29 minutes.

But I agree with you. This design does not preclude a subsequent race, we just need to make sure epochs are timed long enough and then extend the rollup logic.